apps/external_thread_pool_demo/my_thread_pool.hh

```diff
@@ -0,0 +1,135 @@
+/**
+ * DAIL in Alibaba Group
+ *
+ */
+
+#pragma once
+
+#include <seastar/core/internal/poll.hh>
+#include <seastar/core/internal/pollable_fd.hh>
+#include <seastar/core/reactor.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/util/noncopyable_function.hh>
+#include <seastar/util/std-compat.hh>
+#include <boost/lockfree/queue.hpp>
+
+class my_thread_pool;
+
+class work_thread {
+    struct work_item {
+        unsigned _from_shard;
+
+        explicit work_item(unsigned from_shard): _from_shard(from_shard) {}
+        virtual ~work_item() = default;
+        virtual void process() = 0;
+        virtual void complete() = 0;
+    };
+    template <typename T>
+    struct work_item_returning : work_item {
+        seastar::noncopyable_function<T ()> _func;
+        seastar::promise<T> _promise;
+        boost::optional<T> _result{};
+
+        work_item_returning(unsigned from_shard, seastar::noncopyable_function<T ()> func)
+            : work_item(from_shard), _func(std::move(func)) {}
+        void process() override { _result = this->_func(); }
+        void complete() override { _promise.set_value(std::move(*_result)); }
+        seastar::future<T> get_future() { return _promise.get_future(); }
+    };
+    std::unique_ptr<work_item> _work_item;
+    seastar::writeable_eventfd _start_eventfd;
+    seastar::posix_thread _worker;
+    std::atomic<bool> _stopped = { false };
+    seastar::reactor* _employer = nullptr;
+public:
+    ~work_thread();
+private:
+    explicit work_thread(unsigned worker_id);
+
+    template <typename T>
+    seastar::future<T> submit(seastar::reactor* employer, seastar::noncopyable_function<T ()> func) {
+        _employer = employer;
+        auto* wi = new work_item_returning<T>(seastar::this_shard_id(), std::move(func));
+        auto fut = wi->get_future();
+        _work_item.reset(wi);
+        _start_eventfd.signal(1);
+        return fut;
+    }
+
+    void work(unsigned worker_id);
+
+    friend class my_thread_pool;
+};
+
+class my_thread_pool_pollfn;
+
+class my_thread_pool {
+    static unsigned _worker_num;
+    static int64_t _sleep_duration_in_microseconds;
+    static std::unique_ptr<boost::lockfree::queue<work_thread*>> _threads;
+    using completed_queue = boost::lockfree::queue<work_thread::work_item*>;
+    struct completed_queue_deleter {
+        void operator()(completed_queue* q) const;
+    };
+    static std::unique_ptr<completed_queue[], completed_queue_deleter> _completed;
+    struct atomic_flag_deleter {
+        void operator()(std::atomic<bool>* flags) const;
+    };
+    static std::unique_ptr<std::atomic<bool>[], atomic_flag_deleter> _reactors_idle;
+public:
+    static void configure();
+    static void stop();
+
+    template <typename T, typename Func>
+    static seastar::future<T> submit_work(Func func) {
+        return seastar::repeat_until_value([] () -> seastar::future<std::optional<work_thread*>> {
```

Comment:
This is a little cumbersome.  Keep in mind that we futurize the continuations so they can also return values (in all return paths) rather that futurized values.

I think this can be further simplified to:
```c++
    return seastar::repeat_until_value([] {
           work_thread* wt = nullptr;
            if (_threads->pop(wt)) {
                return std::make_optional<work_thread*>(wt);
            }
            return seastar::sleep(std::chrono::microseconds(_sleep_duration_in_microseconds)).then([] {
                return std::nullopt;
            });
    });
```

Reply:
I followed this method but a compilation error occurred. I understand that the code in the upper half returns an optional, but the code in the lower half returns a future.

I tried to modify it as follows, and it works!

```
return seastar::repeat_until_value([] () -> std::optional<work_thread*> {
            work_thread* wt = nullptr;
            if (_threads->pop(wt)) {
                return wt;
            }
            return seastar::sleep(std::chrono::microseconds(_sleep_duration_in_microseconds)).then([] {
                return std::nullopt;
            }).get0();
});
```